reconcile/terraform_repo.py

```diff
@@ -0,0 +1,304 @@
+    """
+    return TFRepo(
+        name=repo.name,
+        repository=repo.repository,
+        ref=repo.ref,
+        project_path=repo.project_path,
+        account=AWSAccount(
+            name=repo.account.name,
+            uid=repo.account.uid,
+            secret=AWSAuthSecret(
+                path=repo.account.automation_token.path,
+                version=repo.account.automation_token.version,
+            ),
+        ),
+        delete=repo.delete,
+    )
+
+
+def merge_results(
+    created: Mapping[str, TFRepo], updated: Mapping[str, TFRepo]
+) -> list[TFRepo]:
+    """
+    Merges results into a RepoOutput dict which will be transformed to outputted YAML.
+    This includes checking modified values for a delete flag
+    """
+    output: list[TFRepo] = []
+    for c_value in created.values():
+        logging.info(["create_repo", c_value.account.name, c_value.name])
+        output.append(c_value)
+    for u_value in updated.values():
+        if u_value.delete is True:
+            logging.info(["delete_repo", u_value.account.name, u_value.name])
+            output.append(u_value)
+        else:
+            logging.info(["update_repo", u_value.account.name, u_value.name])
+            output.append(u_value)
+    return output
+
+
+def update_state(
+    created: Mapping[str, TFRepo],
+    deleted: Mapping[str, TFRepo],
+    updated: Mapping[str, TFRepo],
+    state: State,
+) -> None:
+    """
+    State represents TFRepo data structures similar to their GQL representations.
+    In regards to deleting a Terraform Repo, when the delete flag is set to True, then
+    the state representation of this repo is also deleted even though the definition may
+    still exist in App Interface
+    """
+    created_and_updated = {**created, **updated}
+    try:
+        for cu_key, cu_val in created_and_updated.items():
+            if cu_val.delete is True:
+                state.rm(cu_key)
+            else:
+                state.add(cu_key, cu_val.json(), True)
+        for d_key in deleted.keys():
+            state.rm(d_key)
+    except KeyError:
+        pass
+
+
+def calculate_diff(
+    existing_state: list[TFRepo],
+    desired_state: list[TFRepo],
+    dry_run: bool,
+    state: Optional[State],
+) -> list[TFRepo]:
+    """Diffs existing and desired state as well as updates the state in S3 if this is not a dry-run operation"""
+    existing_map = map_repos(existing_state)
+    desired_map = map_repos(desired_state)
+
+    to_be_created = diff_missing(desired_map, existing_map)
+    to_be_updated = diff_changed(existing_map, desired_map)
+
+    # indicates repos which have had their definitions deleted from App Interface
+    # a pre-requisite to this step is setting the delete flag in the repo definition
+    to_be_deleted = diff_missing(existing_map, desired_map, True)
+
+    if not dry_run and state:
+        update_state(to_be_created, to_be_deleted, to_be_updated, state)
```

Comment:
Instead of having these dictionary comparisons, you may use `set` comparison. e.g https://docs.python.org/3/library/stdtypes.html#frozenset.difference

pseudo code:
```
def calculate ...
existing_state: list[TFRepo],
desired_state: list[TFRepo],

existing = set(existing_state)
desired = set(desired_state)

to_be_created = desired - existing
to_be_deleted = existing - desired

you need to implement a `__eq__` method, e.g. https://github.com/app-sre/qontract-reconcile/blob/master/reconcile/skupper_network/models.py#L39
```

Reply:
@chassing I noticed we are using this diff pattern a lot, also suggested something similar before, would it worth extracting a generic util function to diff? Given two mappings, return a symmetric difference result like

```
   only_in_a: [A]
   only_in_b: [B]
   both: [(A, B)]
```

Reply:
I was also wondering about a generic diff function when writing this heh. Poking around it seems like most integrations take their own approach at performing diffs when it could be consolidated.

Reply:
I created one in https://github.com/app-sre/qontract-reconcile/pull/3475, please see if it's easy to use. If we can try it in this new integration successfully, then we can apply it to others.

Reply:
Works for me!